File: mobile/modules/orca-mobile-web-shell/android/src/main/java/expo/modules/orcamobilewebshell/MobileWebShellLoadState.kt
package expo.modules.orcamobilewebshell

/** The wire names the TypeScript parser accepts; a swap here is a silent change of meaning. */
internal enum class MobileWebShellFailureReason(val wireName: String) {
  GENERATION_UNREADABLE("generation-unreadable"),
  ISOLATION_UNAVAILABLE("isolation-unavailable"),
  DOCUMENT_LOAD_FAILED("document-load-failed"),
  RENDER_PROCESS_GONE("render-process-gone")
}

internal data class MobileWebShellLoadEmission(val state: String, val reason: String?) {
  fun toPayload(): Map<String, Any> = if (reason == null) {
    mapOf("state" to state)
  } else {
    mapOf("state" to state, "reason" to reason)
  }
}

/**
 * What a mount is still allowed to report. A failure is terminal: Chromium commits its own error
 * document after `onReceivedError` returns, and a rule list can fail to compile long after the
 * generation was already refused, so without this a `ready` or a second reason lands on top of a
 * failure the caller has already acted on. Consecutive duplicates are dropped as well.
 *
 * Pure, and the same rule on both platforms, so a JVM test and a `swiftc` check can hold it. The
 * two fields a caller reads directly are volatile: Android decides a document failure from
 * `shouldInterceptRequest`, which Chromium does not run on the UI thread.
 */
internal class MobileWebShellLoadStateMachine {
  private var terminal = false
  private var last: MobileWebShellLoadEmission? = null

  /** Which load this machine is reporting on. Read before deferring work, checked on delivery. */
  @Volatile
  var epoch: Int = 0
    private set

  /**
   * Whether a document under the current prop triple has committed. The document a load replaces
   * stays alive between `stopLoading` and the next commit, and it is same-origin whenever only the
   * directory or the bridge prop changed, so without this it passes every origin check and speaks
   * for a load the caller has already been told is `loading`.
   */
  @Volatile
  var hasCommittedDocument = false
    private set

  /** A new prop pair. Nothing else reopens a terminal state: a retry is a remount. */
  fun reset() {
    terminal = false
    last = null
    epoch += 1
    documentEnded()
  }

  fun committed() {
    if (terminal) return
    hasCommittedDocument = true
  }

  /** The committed document is gone: a new load, a failure, or a renderer that died. */
  fun documentEnded() {
    hasCommittedDocument = false
  }

  fun started(): MobileWebShellLoadEmission? = emit(MobileWebShellLoadEmission("loading", null))

  /**
   * A load that never committed did not finish.
   *
   * This is the whole guard, and it is deliberately not the document's URL: the page rewrites its
   * own path with `history.replaceState` before its first render, so the document that committed at
   * "/" reports finishing at "/h/<hostId>". Reading the path here withheld `ready` forever and left
   * the WebView hidden behind it.
   */
  fun finished(): MobileWebShellLoadEmission? {
    if (!hasCommittedDocument) return null
    return emit(MobileWebShellLoadEmission("ready", null))
  }

  fun failed(reason: MobileWebShellFailureReason): MobileWebShellLoadEmission? {
    val emission = emit(MobileWebShellLoadEmission("failed", reason.wireName))
    terminal = true
    documentEnded()
    return emission
  }

  /**
   * A failure decided during one load and reported after the next one started belongs to neither:
   * Android has to defer its report past Chromium's error document, and a prop update can land in
   * between, which would fail the generation that just replaced the one that actually failed.
   */
  fun failedDuring(epoch: Int, reason: MobileWebShellFailureReason): MobileWebShellLoadEmission? =
    if (epoch != this.epoch) null else failed(reason)

  private fun emit(emission: MobileWebShellLoadEmission): MobileWebShellLoadEmission? {
    if (terminal || emission == last) return null
    last = emission
    return emission
  }
}

File: mobile/modules/orca-mobile-web-shell/ios/MobileWebShellLoadState.swift
import Foundation

/// The wire names the TypeScript parser accepts; a swap here is a silent change of meaning.
enum MobileWebShellFailureReason: String {
  case generationUnreadable = "generation-unreadable"
  case isolationUnavailable = "isolation-unavailable"
  case documentLoadFailed = "document-load-failed"
  case renderProcessGone = "render-process-gone"
}

struct MobileWebShellLoadEmission: Equatable {
  let state: String
  let reason: String?
}

/// What a mount is still allowed to report. A failure is terminal: a rule list can fail to compile
/// long after the generation was already refused, and WebKit still reports a navigation outcome
/// after a response was cancelled, so without this a second reason or a `ready` lands on top of a
/// failure the caller has already acted on. Consecutive duplicates are dropped as well.
///
/// Pure, and the same rule as the Kotlin copy, so `swiftc` can check it without a device.
final class MobileWebShellLoadStateMachine {
  private var isTerminal = false
  private var last: MobileWebShellLoadEmission?

  /// Whether a document under the current prop triple has committed. The document a load replaces
  /// stays alive between `stopLoading` and the next commit, and it is same-origin whenever only the
  /// directory or the bridge prop changed, so without this it passes every origin check and speaks
  /// for a load the caller has already been told is `loading`.
  private(set) var hasCommittedDocument = false

  /// Whether the navigation in flight is the one the shell asked for.
  ///
  /// Kept here rather than beside the `load` call because every way a document can end already runs
  /// through this type: a commit, a failure, a renderer that died, a prop update that never loaded.
  /// A flag in the view had to remember each of those separately, and missed two.
  private(set) var isShellLoad = false

  /// The view is about to load the document itself. The only thing that raises the flag.
  func shellLoadStarted() {
    isShellLoad = true
  }

  /// The one navigation the flag was raised for has been allowed, so the flag is spent.
  ///
  /// Spent at the decision and not at the commit: WebKit can decide a second main-frame action
  /// before the first one starts, and a flag still raised then would have allowed that one to
  /// replace the document.
  func shellLoadConsumed() {
    isShellLoad = false
  }

  /// A new prop pair. Nothing else reopens a terminal state: a retry is a remount.
  func reset() {
    isTerminal = false
    last = nil
    documentEnded()
  }

  func committed() {
    isShellLoad = false
    guard !isTerminal else { return }
    hasCommittedDocument = true
  }

  /// The committed document is gone: a new load, a failure, or a renderer that died.
  func documentEnded() {
    hasCommittedDocument = false
    isShellLoad = false
  }

  func started() -> MobileWebShellLoadEmission? {
    emit(MobileWebShellLoadEmission(state: "loading", reason: nil))
  }

  /// A load that never committed did not finish.
  ///
  /// This is the whole guard, and it is deliberately not the document's URL: the page rewrites its
  /// own path with `history.replaceState` before its first render, so the document that committed
  /// at "/" reports finishing at "/h/<hostId>". Reading the path here withheld `ready` forever.
  func finished() -> MobileWebShellLoadEmission? {
    guard hasCommittedDocument else { return nil }
    return emit(MobileWebShellLoadEmission(state: "ready", reason: nil))
  }

  func failed(_ reason: MobileWebShellFailureReason) -> MobileWebShellLoadEmission? {
    let emission = emit(MobileWebShellLoadEmission(state: "failed", reason: reason.rawValue))
    isTerminal = true
    documentEnded()
    return emission
  }

  private func emit(_ emission: MobileWebShellLoadEmission) -> MobileWebShellLoadEmission? {
    guard !isTerminal, emission != last else { return nil }
    last = emission
    return emission
  }
}

/// A navigation WebKit reports as failed but which is not a failure of the document.
///
/// `stopLoading` on a prop update, and every navigation the policy delegate refuses, arrive at the
/// failure delegates as errors. Reporting those would fail a healthy page, swallow its `ready`, and
/// send the caller off to delete a cached generation that is fine.
///
/// The WebKit constant is written out because the iOS SDK exports no symbol for it: `WKErrorCode`
/// stops at the content-rule-list and app-bound-domain errors, and the frame-load codes live in the
/// legacy `WebKitErrorDomain`, which WKWebView still reports a policy-cancelled frame load under.
enum MobileWebShellNavigationError {
  static let webKitDomain = "WebKitErrorDomain"
  static let frameLoadInterruptedByPolicyChange = 102

  static func isIgnorable(domain: String, code: Int) -> Bool {
    if domain == NSURLErrorDomain, code == NSURLErrorCancelled {
      return true
    }
    return domain == webKitDomain && code == frameLoadInterruptedByPolicyChange
  }
}
